### DISCUSSION TITLE
RustDesk behind Traefik

### DISCUSSION THREAD

**Discussion by Nomtest** [rustdesk/rustdesk#3177] [Q&A]
Hello,

How to get  RustDesk behind Traefik ?

Is it so possible to get a link in a webpage to allow user to begin a RustDesk session with me ?

Tnx !

**Comment by saitoh183** [comment]
I would be interested in this also

**Comment by FracKenA** [comment]
I have it working... ish.

Below is a sample of a combined traefik/rustdesk/whoami docker-compose. 

https://gist.github.com/FracKenA/5015150a6057afe9a22bf2e4b9d5f2b6

It currently works inside my firewall, it doesn't work outside though. Traefik gets the following error.

```
time="2023-03-14T18:41:22Z" level=error msg="Error while handling UDP stream: read udp 172.19.0.4:43242->172.19.0.2:21116: read: connection refused"
```

I saw this previously when I wasn't attempting to use traefik and I had to specify the relay server in my clients to resolve, however now it is no longer working when behind traefik.

**Reply by saitoh183** [reply]
I gave up traefik and moved to Caddy but ultimately i just opened the ports for it in my Firewall less of a hassle

**Reply by FracKenA** [reply]
Totally get it, if you decide to try it again, the above linked gist works now.

**Comment by FracKenA** [comment]
The above challenge is now resolved. Working using tcp and udp through traefik.

**Comment by hadim** [comment]
For the record, see this configuration as well https://gist.github.com/hadim/2225f0869ada523dec8148510ca438ca

**Comment by xlionjuan** [comment]
Won't increase security unless you implement some ACL, but you could implement it in your firewall.....and more easy

**Reply by FracKenA** [reply]
Correct. 

I This is not intended as a security solution, this is simply putting it behind the load balancer. Security should always be taken into consideration using a viable solution like a firewall.

**Comment by TBG-FR** [comment]
Working Traefik configuration for HTTPS RustDesk Web Client 😉 https://github.com/rustdesk/rustdesk/discussions/6023#discussioncomment-8687562

**Comment by rpgdev** [comment]
> Working Traefik configuration for HTTPS RustDesk Web Client 😉 [#6023 (reply in thread)](https://github.com/rustdesk/rustdesk/discussions/6023#discussioncomment-8687562)

What exactly is it accomplishing? Do you still need to vpn? Or were you able to avoid exposing the rustdesk ports beyond your firewall? Have you shared the entire docker-compose.yml somewhere?

**Reply by TBG-FR** [reply]
You don't need a VPN to use RustDesk, wether you use Desktop, Mobile or Web client.

However, you do need to expose RustDesk ports on the server side (hbbs, hbbr) in order to make it work, that's the expected behaviour of any server-side component (exposing server container ports directly, or exposing them through Traefik)

The setup I shared "only" allows to expose the Web client on a HTTPS endpoint, and make a WSS connection between the Web client and the reverse proxy. After that, the reverse proxy (Traefik) will internally make a insecure connection to the server's containers (because the server doesn't support WSS https://github.com/rustdesk/rustdesk-server/issues/94 https://github.com/rustdesk/rustdesk-server/issues/337) but we don't care, it's internal and not exposed

**Comment by rpgdev** [comment]
> However, you do need to expose RustDesk ports on the server side (hbbs, hbbr) in order to make it work, that's the expected behaviour of any server-side component (exposing server container ports directly, or exposing them through Traefik)
> 
> The setup I shared "only" allows to expose the Web client on a HTTPS endpoint, and make a WSS connection between the Web client and the reverse proxy. After that, the reverse proxy (Traefik) will internally make a insecure connection to the server's containers (because the server doesn't support WSS [rustdesk/rustdesk-server#94](https://github.com/rustdesk/rustdesk-server/issues/94) [rustdesk/rustdesk-server#337](https://github.com/rustdesk/rustdesk-server/issues/337)) but we don't care, it's internal and not exposed

So with this setup you can't rustdesk from beyond the firewall with any other client besides the web client, correct?

**Reply by TBG-FR** [reply]
It changes nothing from the non HTTPS setup for that matter, if you want a client (web or anything else) to connect to a server (RustDesk or anything else), the server should be exposed through ports. Firewall on client side should not be an issue, firewall on server side however should allow communications on these ports

**Comment by rpgdev** [comment]
> It changes nothing from the non HTTPS setup for that matter, if you want a client (web or anything else) to connect to a server (RustDesk or anything else), the server should be exposed through ports. Firewall on client side should not be an issue, firewall on server side however should allow communications on these ports

So I have to open the ports through the firewall regardless? My server would need to expose those ports (or forwarded ports that map to them) over the internet? Just want to make sure what it entails, I would've wanted to avoid that if possible!

**Reply by TBG-FR** [reply]
Yup, that's how it works 😉
- If you're able to `ssh` to a server, it's because its (default 22) port is open
- If you're able to get a response (web page) on https://google.com, it's because there's a server exposing ports 80 (HTTP) and 443 (HTTPS)

So
- RustDesk server has to expose its ports (21115, 21116, 21117, 21118, 21119) to be reachable by clients
- RustDesk web client has to be exposed somewhere too, the same way as any web page/app

The goal of Traefik (or any other reverse proxy) is to put multiple services/containers/webpages without exposing them directly, everything goes through Traefik and then it redirects to your services **internally**, based on rules . That's why you see in Traefik setups
- `ports:` Docker instructions, to expose ports
- `--entrypoints` Traefik commands, to define entrypoints matching these ports

Usually on a regular provided linux server, you won't even have to modify your firewall rules, docker/traefik will do it. But yes if you have strict firewall rules (or company policy enforcing them), you'll need to open ports in the firewall

**Reply by rpgdev** [reply]
Ok, I 'know' that just wanted to confirm if you had found an alternative using reverse proxies (by assigning urls that map to different ports or something like that). I use nginx proxy manager so I'm not that well versed on traefik and didn't understand your setup at all. Will probably need to start using traefik since it does have some benefits over nginx on docker. Thanks for taking the time to respond!
